doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,487 @@
+_at least_ up to date with the catalog state as of which the peek was
+processed. Note that we say the Controller has to be _at least_ up to date with
+the `catalog_as_of`, it is okay for its catalog frontier to be beyond that.
+This has the effect that in-flight peeks can end up in a situation where a
+Controller doesn't have the collection that we are trying to query, and we
+therefore have to report back an error. This is in line with current
+Materialize behavior.
+
+As we see below in the section on ADAPTER. We replace explicit sequencing by
+decoupling using pTVCs and timestamps.
+
+- `Peek(global_id, catalog_as_of, ..)`: performs the peek, but _only_ once the
+  catalog frontier has been downgraded far enough.
+
+  It is important to note that the new `catalog_as_of` parameter is different
+  from an `as_of`, which latter refers to the timeline/timestamp of the object
+  being queried.
+
+- `AdvanceCatalogFrontier(timestamp): tells the controller that it has seen
+  commands corresponding to catalog changes up to `timestamp`.
+
+(Most existing commands have been omitted for brevity!)
+
+### ADAPTER
+
+ADAPTER spawns threads/processes/tasks for handling client connections as they
+show up.
+
+Additionally, it spawns one task per controller that runs a loop that
+synthesizes commands for the controller based on differential CATALOG changes.
+To do this, the task subscribes to CATALOG, filters on changes that are
+relevant for the given controller, sends commands to the controller while
+preserving the order in which they are observed in the stream of changes, and
+synthesizes `AdvanceCatalogFrontier` commands when the catalog timestamp
+advances. These commands are the commands that deterministically derive from
+CATALOG state (see Background section, above).
+
+When processing client requests, ADAPTER uses TIMESTAMP ORACLE to get the
+latest read timestamp for CATALOG, then fetches a CATALOG snapshot as of at
+least that timestamp, and continues to process the request using the catalog
+snapshot as the source of truth.
+
+When handling DDL-style client queries, ADAPTER uses TIMESTAMP ORACLE to get a
+write timestamp and tries to append the required changes to the CATALOG. If
+there is a conflict it has to get a new timestamp, get the latest CATALOG
+snapshot, and try again. ADAPTER does _not_ explicitly instruct the controllers
+to do anything that is required by these changes and it does _not_ wait for
+controllers to act on those changes before returning to the client. Controllers
+are expected to learn about these changes and act upon them from their
+"private" command loop.
+
+Client queries that spawn ephemeral controller commands (mostly DQL and
+DML-style queries) are timestamped with the timestamp (in the catalog timeline)
+as of which the query was processed. This ensure that these ephemeral
+controller commands are consistent with what the controller must know as of
+that timestamp.
```

Comment:
One thing I don't see discussed here is the nondeterminism that can be _caused_ by controller commands. Specifically thinking of how the storage controller today is responsible for minting shard IDs. Is this the moment where we want to declare that everything the controller does should be a perfectly deterministic function of its inputs?

Reply:
I'm 100% of the opinion that having one durable store, that you write to atomically is easier to reason about than having multiple of those (the current situation). And think this might be required to do 0dt upgrades.

I'm less certain we "really" need it here already: the controller makes the `GlobalId -> ShardId` mapping definite before it returns from an operation.

This assumes, though, that the mapping (or anything else really) that the controller writes down cannot change. I feel this is similar to what you outlined in https://github.com/MaterializeInc/materialize/pull/23543#discussion_r1435916930. Then again, though, it seems we also need a revision-like mechanism if we ever want to allow changing the mapping (or any other controller state we keep in shared durable storage).